### PR TITLE
macOS: Add a `bsd_flags` column to the `file` table

### PR DESCRIPTION
**Pull request by GarretReece**
Initial work by @alessandrogario , git wrangling to get it playing with current master by @GarretReece

This PR implements support for the BSD flags on macOS as documented in issue #4189.

```console
garretreece@latke-2:~$ ./osqueryd -S "select path, bsd_flags from file where path='/Users/garretreece/test-file'"
+------------------------------+-----------+
| path                         | bsd_flags |
+------------------------------+-----------+
| /Users/garretreece/test-file |           |
+------------------------------+-----------+
garretreece@latke-2:~$ chflags nodump,hidden test-file 
garretreece@latke-2:~$ ./osqueryd -S "select path, bsd_flags from file where path='/Users/garretreece/test-file'"
+------------------------------+----------------+
| path                         | bsd_flags      |
+------------------------------+----------------+
| /Users/garretreece/test-file | HIDDEN, NODUMP |
+------------------------------+----------------+
garretreece@latke-2:~$
```

### CODE REVIEW

**Comment by theopolis** [comment]
IMO if we can avoid pre-processor logic we should, and instead use the platform-testing APIs in `./utils`.

File: osquery/core/darwin/bsd_file_flags.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
+#include <iomanip>
```
Comment (theopolis) [inline osquery/core/darwin/bsd_file_flags.cpp]: Do you mind adding the copyright header?

File: osquery/tables/utility/file.cpp [hunk 2]
```diff
@@ -97,6 +97,17 @@ void genFileInfo(const fs::path& path,
     r["type"] = "unknown";
   }
 
+#if defined(__APPLE__)
+  std::string bsd_file_flags_description;
+  if (!describeBSDFileFlags(bsd_file_flags_description, file_stat.st_flags)) {
+    LOG(WARNING)
```
Comment (theopolis) [inline osquery/tables/utility/file.cpp]: It seems like this should be a `VLOG(1)` instead of a warning?

File: osquery/filesystem/fileops.h [hunk 3]
```diff
@@ -426,4 +426,8 @@ boost::filesystem::path getSystemRoot();
  * @return osquery::Status
  */
 Status platformLstat(const std::string& path, struct stat& d_stat);
+
+#ifdef __APPLE__
+Status describeBSDFileFlags(std::string& output, std::uint32_t st_flags);
```
Comment (theopolis) [inline osquery/filesystem/fileops.h]: Can you add a docblock for the function?

File: osquery/core/tests/darwin/bsd_file_flags_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
+#include <unordered_map>
```
Comment (theopolis) [inline osquery/core/tests/darwin/bsd_file_flags_tests.cpp]: Same, w.r.t. header.

File: osquery/filesystem/fileops.h [hunk 5]
```diff
@@ -426,4 +426,8 @@ boost::filesystem::path getSystemRoot();
  * @return osquery::Status
  */
 Status platformLstat(const std::string& path, struct stat& d_stat);
+
+#ifdef __APPLE__
```
Comment (theopolis) [inline osquery/filesystem/fileops.h]: My recommendation is to drop this check, allow the function to exist on all platforms.

File: osquery/tables/utility/file.cpp [hunk 6]
```diff
@@ -97,6 +97,17 @@ void genFileInfo(const fs::path& path,
     r["type"] = "unknown";
   }
 
+#if defined(__APPLE__)
```
Comment (theopolis) [inline osquery/tables/utility/file.cpp]: My recommendation is to drop this, instead gate this logic with `isPlatform(PlatformType::TYPE_OSX)`

Comment (GarretReece) [reply osquery/tables/utility/file.cpp]: I'd like to, but we get an error on non-darwin platforms:
```console
osquery/tables/utility/file.cpp:103:69: error: no member named 'st_flags' in 'stat'
    if (!describeBSDFileFlags(bsd_file_flags_description, file_stat.st_flags)) {
                                                          ~~~~~~~~~ ^
```

**Comment by theopolis** [comment]
I just noticed that files are being added to `./osquery/core`. I think we should avoid this unless the code is very internal to osquery's operation. I suggest placing this within the `./osquery/filesystem` abstraction or alongside table implementation code within `./osquery/tables`.

File: osquery/filesystem/darwin/bsd_file_flags.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+} // namespace
+
+Status describeBSDFileFlags(std::string& output, std::uint32_t st_flags) {
+  output.clear();
+
+  static const auto flag_mask = getBsdFlagMask();
+
+  std::vector<std::string> label_list;
+
+  for (const auto& p : kBsdFlagMap) {
```
Comment (directionless) [inline osquery/filesystem/darwin/bsd_file_flags.cpp]: Will this always iterate in the same order? I think it's desirable for `output` to have a stable order.

File: osquery/filesystem/BUCK [hunk 8]
```diff
@@ -104,7 +110,8 @@ osquery_cxx_test(
     platform_srcs = [
         (
             MACOSX,
-            ["tests/darwin/plist_tests.cpp"],
+            ["tests/darwin/plist_tests.cpp",
```
Comment (theopolis) [inline osquery/filesystem/BUCK]: Can you write this as:
```
            [
                "tests/darwin/plist_tests.cpp",
                "tests/darwin/bsd_file_flags_tests.cpp"
            ],

File: osquery/filesystem/CMakeLists.txt [hunk 9]
```diff
@@ -101,7 +107,10 @@ function(generateOsqueryFilesystemTest)
   )
 
   if(DEFINED PLATFORM_MACOS)
-    list(APPEND source_files tests/darwin/plist_tests.cpp)
+    list(APPEND source_files 
+        tests/darwin/plist_tests.cpp
```
Comment (theopolis) [inline osquery/filesystem/CMakeLists.txt]: The indent is wrong here, it should be:
```
    list(APPEND source_files 
      tests/darwin/plist_tests.cpp
      tests/darwin/bsd_file_flags_tests.cpp
    )
```

**Pull request merged**